lib/api/bucketGetEncryption.js

```diff
@@ -0,0 +1,84 @@
+const { errors, s3middleware } = require('arsenal');
+
+const bucketShield = require('./apiUtils/bucket/bucketShield');
+const { isBucketAuthorized } =
+    require('./apiUtils/authorization/permissionChecks');
+const metadata = require('../metadata/wrapper');
+const { pushMetric } = require('../utapi/utilities');
+const escapeForXml = s3middleware.escapeForXml;
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+
+const requestType = 'bucketGetEncryption';
+
+/**
+ * Bucket Get Encryption - Get bucket SSE configuration
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+
+function bucketGetEncryption(authInfo, request, log, callback) {
+    const bucketName = request.bucketName;
+    const canonicalID = authInfo.getCanonicalID();
+
+    return metadata.getBucket(bucketName, log, (err, bucket) => {
+        if (err) {
+            log.debug('metadata getbucket failed', { error: err });
+            return callback(err);
+        }
+        if (bucketShield(bucket, requestType)) {
+            return callback(errors.NoSuchBucket);
+        }
+        log.trace('found bucket in metadata');
+
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+
+        if (!isBucketAuthorized(bucket, requestType, canonicalID, authInfo,
+        log)) {
+            log.debug('access denied for account on bucket', {
+                requestType,
+                method: 'bucketGetEncryption',
+            });
+            return callback(errors.AccessDenied, null, corsHeaders);
+        }
+
+        const sseInfo = bucket.getServerSideEncryption();
+
+        // If ssInfo is present but the `mandatory` flag is not set
+        // then this info was not created using bucketPutEncryption
+        // and should not be returned
+        if (sseInfo === null || !sseInfo.mandatory) {
```

Comment:
In the design doc, I suggested we could return the existing "legacy" encryption config via GetBucketEncryption. Not sure if it's a good idea or not but let's discuss it and make sure we agree on what to do here

Reply:
As I understand it bucket created using the `x-amz-scal-server-side-encryption` header are in essence the same as a bucket with a default scheme defined using `PutBucketEncryption`

Reply:
Yes, those should be compatible regarding metadata, I was thinking of using the `mandatory` flag as a way to know if the policy is applied to the bucket (so either via PutBucketEncryption or via the legacy x-scal header) or if it only contains the default encryption params used to encrypt objects that explicitly contain the `x-amx-server-side-encryption` header, i.e. the master key ID and cryptoScheme, since ideally a bucket always uses the same master key by default (except if specified explicitly). I cannot tell for sure what was the original intent of `mandatory`, but I think it is somehow in the same spirit, except that without the backward-compat thing I'd rather call it `applyByDefault` but we have to keep the name. This way we can also support object-level encryption with minimal changes to the metadata format. I can change this if it does not fit well.